test/dlc_tests/sanity/test_security_check.py

```diff
@@ -54,6 +66,297 @@ def run_scan(ecr_client, image):
         raise TimeoutError(f"ECR Scan is still in {scan_status} state. Exiting.")
 
 
+def get_new_image_uri(image):
+    """
+    Returns the new image uri for the image being tested. After running the apt commands, the
+    new image will be uploaded to the ECR based on the new image uri.
+
+    :param image: str
+    :param new_image_uri: str
+    """
+    repository_name = os.getenv("UPGRADE_REPO_NAME")
+    ecr_account = f"{os.getenv('ACCOUNT_ID')}.dkr.ecr.{os.getenv('REGION')}.amazonaws.com"
+    upgraded_image_tag = "-".join(image.replace("/", ":").split(":")[1:]) + "-up"
+    new_image_uri = f"{ecr_account}/{repository_name}:{upgraded_image_tag}"
+    return new_image_uri
+
+
+def run_upgrade_on_image_and_push(image, new_image_uri):
+    """
+    Creates a container for the image being tested. Runs apt update and upgrade on the container
+    and the commits the container as new_image_uri. This new image is then pushed to the ECR. 
+
+    :param image: str
+    :param new_image_uri: str
+    """
+    ctx = Context()
+    docker_run_cmd = f"docker run -id --entrypoint='/bin/bash' {image}"
+    container_id = ctx.run(f"{docker_run_cmd}", hide=True, warn=True).stdout.strip()
+    apt_command = "apt-get update && apt-get upgrade"
+    docker_exec_cmd = f"docker exec -i {container_id}"
+    attempt_count = 0
+    apt_ran_successfully_flag = True
+    while True:
+        run_output = ctx.run(f"{docker_exec_cmd} {apt_command}", hide=True, warn=True)
+        attempt_count += 1
+        if not run_output.ok:
+            LOGGER.info(
+                f"Attempt no. {attempt_count} on image: {image}"
+                f"Could not run apt update and upgrade. \n"
+                f"Stdout is {run_output.stdout} \n"
+                f"Stderr is {run_output.stderr} \n"
+                f"Failed status is {run_output.exited}"
+            )
+            sleep(2 * 60)
+        elif run_output.ok:
+            break
+        if attempt_count == 10:
+            if not run_output.ok:
+                apt_ran_successfully_flag = False
```

Comment:
Why does this need to be attempted repeatedly?

Reply:
When a command or application is updating the system or installing a new software (i.e. using apt get update), it locks the dpkg file (Debian package manager). Since we have multiple processes running for the tests, there are cases when one of the process locks the dpkg file. In this scenario, we get error: ‘E: Could not get lock /var/lib/dpkg/lock’ while running apt-get update. That is why we need multiple tries to ensure that it succeeds in one of the tries. More info is seen here: https://itsfoss.com/could-not-get-lock-error/